internal/provider/resource_password.go

```diff
@@ -1,22 +1,109 @@
 package provider
 
 import (
+	"context"
+	"fmt"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"golang.org/x/crypto/bcrypt"
 )
 
 func resourcePassword() *schema.Resource {
+	create := func(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+		diags := createStringFunc(true)(ctx, d, meta)
+		if diags.HasError() {
+			return diags
+		}
+
+		hash, err := generateHash(d.Get("result").(string))
+		if err != nil {
+			diags = append(diags, diag.Errorf("err: %s", err)...)
+			return diags
+		}
+
+		if err := d.Set("bcrypt_hash", hash); err != nil {
+			diags = append(diags, diag.Errorf("err: %s", err)...)
+			return diags
+		}
+
+		return nil
+	}
+
 	return &schema.Resource{
 		Description: "Identical to [random_string](string.html) with the exception that the result is " +
 			"treated as sensitive and, thus, _not_ displayed in console output. Read more about sensitive " +
 			"data handling in the [Terraform documentation](https://www.terraform.io/docs/language/state/sensitive-data.html).\n" +
 			"\n" +
 			"This resource *does* use a cryptographic random number generator.",
-		CreateContext: createStringFunc(true),
+		CreateContext: create,
 		ReadContext:   readNil,
 		DeleteContext: RemoveResourceFromState,
-		Schema:        stringSchemaV1(true),
+		Schema:        passwordSchemaV1(),
 		Importer: &schema.ResourceImporter{
-			StateContext: importStringFunc(true),
+			StateContext: importPasswordFunc(),
 		},
+		SchemaVersion: 1,
+		StateUpgraders: []schema.StateUpgrader{
+			{
+				Version: 0,
+				Type:    resourcePasswordV0().CoreConfigSchema().ImpliedType(),
+				Upgrade: resourcePasswordStateUpgradeV0,
+			},
+		},
+	}
+}
+
+func importPasswordFunc() schema.StateContextFunc {
```

Comment:
Why do you need a function that returns a function here?

Reply:
Have removed.